File: Gemini_Pro/Chain-of-responsibilityPattern/Round1/Code.py
from abc import ABC, abstractmethod

class Handler(ABC):
    """
    Base class for handlers in the chain of responsibility.
    """
    _next_handler = None

    def set_next(self, handler):
        """
        Sets the next handler in the chain.
        """
        self._next_handler = handler

    def handle(self, request):
        """
        Handles the request.
        """
        if self.can_handle(request):
            self.process(request)
        elif self._next_handler:
            self._next_handler.handle(request)

    @abstractmethod
    def can_handle(self, request):
        """
        Checks if the handler can process the request.
        """
        pass

    @abstractmethod
    def process(self, request):
        """
        Processes the request.
        """
        pass

class LocationHandler(Handler):
    """
    Checks if the package can be delivered to the specified location.
    """
    def can_handle(self, request):
        return request['location'] == 'Valid Location'

    def process(self, request):
        print(f"LocationHandler: Package can be delivered to {request['location']}.")

class PaymentHandler(Handler):
    """
    Handles payment processing for the delivery.
    """
    def can_handle(self, request):
        return request['payment_method'] == 'Valid Payment'

    def process(self, request):
        print("PaymentHandler: Payment processed successfully.")

class DeliveryHandler(Handler):
    """
    Manages the actual delivery of the package.
    """
    def can_handle(self, request):
        return True

    def process(self, request):
        print("DeliveryHandler: Package delivered successfully.")

def main():
    # Create the chain of handlers.
    location_handler = LocationHandler()
    payment_handler = PaymentHandler()
    delivery_handler = DeliveryHandler()

    # Link the handlers together.
    location_handler.set_next(payment_handler)
    payment_handler.set_next(delivery_handler)

    # Create a request.
    request = {
        'location': 'Valid Location',
        'payment_method': 'Valid Payment',
    }

    # Process the request.
    location_handler.handle(request)

if __name__ == '__main__':
    main()

File: Java_ChatGPT/Chain_of_responsbility/Round3/Code.java
package Round3;

// DeliveryRequest class that holds details about the package
class DeliveryRequest {
    private String location;
    private boolean paymentProcessed;
    private boolean isDelivered;

    public DeliveryRequest(String location) {
        this.location = location;
        this.paymentProcessed = false;
        this.isDelivered = false;
    }

    public String getLocation() {
        return location;
    }

    public boolean isPaymentProcessed() {
        return paymentProcessed;
    }

    public void setPaymentProcessed(boolean paymentProcessed) {
        this.paymentProcessed = paymentProcessed;
    }

    public boolean isDelivered() {
        return isDelivered;
    }

    public void setDelivered(boolean delivered) {
        isDelivered = delivered;
    }
}

// Abstract Handler in the chain
abstract class Handler {
    protected Handler nextHandler;

    public void setNextHandler(Handler nextHandler) {
        this.nextHandler = nextHandler;
    }

    public abstract void handleRequest(DeliveryRequest request);
}

// LocationHandler verifies if the package can be delivered to the specified location
class LocationHandler extends Handler {
    private static final String VALID_LOCATION = "ValidLocation";

    @Override
    public void handleRequest(DeliveryRequest request) {
        if (VALID_LOCATION.equals(request.getLocation())) {
            System.out.println("Location is valid.");
            if (nextHandler != null) {
                nextHandler.handleRequest(request);
            }
        } else {
            System.out.println("Invalid location. Stopping the chain.");
        }
    }
}

// PaymentHandler handles payment processing
class PaymentHandler extends Handler {
    @Override
    public void handleRequest(DeliveryRequest request) {
        if (!request.isPaymentProcessed()) {
            System.out.println("Processing payment...");
            request.setPaymentProcessed(true);
            if (nextHandler != null) {
                nextHandler.handleRequest(request);
            }
        } else {
            System.out.println("Payment already processed. Stopping the chain.");
        }
    }
}

// DeliveryHandler manages the actual delivery of the package
class DeliveryHandler extends Handler {
    @Override
    public void handleRequest(DeliveryRequest request) {
        if (request.isPaymentProcessed()) {
            System.out.println("Delivering package...");
            request.setDelivered(true);
        } else {
            System.out.println("Cannot deliver package. Payment not processed.");
        }
    }
}

// Chain of Responsibility setup
public class Code {
    public static void main(String[] args) {
        // Create handlers
        Handler locationHandler = new LocationHandler();
        Handler paymentHandler = new PaymentHandler();
        Handler deliveryHandler = new DeliveryHandler();

        // Set up the chain
        locationHandler.setNextHandler(paymentHandler);
        paymentHandler.setNextHandler(deliveryHandler);

        // Create a valid delivery request
        DeliveryRequest request = new DeliveryRequest("ValidLocation");
        locationHandler.handleRequest(request);

        // Create an invalid delivery request
        DeliveryRequest invalidRequest = new DeliveryRequest("InvalidLocation");
        locationHandler.handleRequest(invalidRequest);
    }
}

